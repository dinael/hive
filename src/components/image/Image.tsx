import { FC } from 'react'
import styles from './Image.module.scss'

export type ImageProps = {
  src: string
  srcSet?: { src: string; size: string }[]
  alt: string
  width?: string
  height?: string
  ratio?: 'none' | 'ratio-1:1' | 'ratio-1:2' | 'ratio-2:1' | 'ratio-2:3' | 'ratio-3:1' | 'ratio-3:2' | 'ratio-3:4' | 'ratio-4:3' | 'ratio-4:5' | 'ratio-5:4' | 'ratio-9:16' | 'ratio-16:9' | 'ratio-16:9' | 'ratio-21:9' | 'ratio-72:25'
  fit?: 'contain' | 'cover' | 'fill' | 'none' | 'scale-down' | 'unset'
  disabled?: boolean
  className?: string
}

export const Image: FC<ImageProps> = ({
  src,
  srcSet,
  alt,
  width = undefined,
  height = undefined,
  ratio = 'none',
  fit = 'cover',
  disabled = false,
  className,
  ...props
}: ImageProps) => {
  if (!srcSet || !Array.isArray(srcSet) || srcSet.length === 0) {
    return (
      <img
        className={`
          ${styles['image-component']}
          ${disabled === true ? styles['disabled'] : ''}
          ${ratio && styles[ratio]}
          ${fit && styles[fit]}
          ${className ? className : ''}
        `}
        src={src}
        alt={alt}
        width={width}
        height={height}
        loading="lazy"
        {...props} />
    )
  }

  const imageDefault = srcSet[0].src

  const srcSetString = srcSet
    .map((obj) => `${obj.src} ${obj.size}`).join(',')
// TODO: Add responsive sizes into sizes parameter
  return (
    <img
      className={`
        ${styles['image-component']}
        ${disabled === true ? styles['disabled'] : ''}
        ${ratio && styles[ratio]}
        ${fit && styles[fit]}
        ${className && className}
      `}
      src={imageDefault}
      srcSet={srcSetString}
      alt={alt}
      width={width ? width : ''}
      height={height ? height : ''}
      loading="lazy"
      {...props} />
  )
}

export default Image


