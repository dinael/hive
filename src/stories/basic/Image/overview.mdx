import { Canvas, Meta } from "@storybook/blocks";

<Meta title="Components/Basic/Image/Overview" />

# v0.0.11 (Wed May 31 2023)
