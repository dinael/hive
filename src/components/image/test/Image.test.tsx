import { describe, test, expect } from 'vitest'
import { render } from '@testing-library/react'
import { Image, ImageProps } from '../Image'

describe('Image', () => {
  const defaultProps: ImageProps = {
    src: 'https://picsum.photos/1440/500?random=1',
    alt: 'Imagen description for accessibility',
    width: '300',
    height: '300',
    ratio: 'none',
    fit: 'cover',
    disabled: true,
    className: 'className-test',
  }

  const imageSet = [
    { src: 'https://picsum.photos/1440/500?random=1', size:'320w' },
    { src: 'https://picsum.photos/900/300?random=2', size:'480w' },
    { src: 'https://picsum.photos/800/200?random=3', size:'640w' },
  ]

  test('renders snapshot', () => {
    const { container } = render(<Image {...defaultProps} />)
    expect(container).toMatchSnapshot()
  })

  test('renders correctly', () => {
    const { container } = render(<Image {...defaultProps} />)
    const imageElement = container.querySelector('img')
    expect(imageElement?.tagName).toBe('IMG')
    expect(imageElement?.getAttribute('src')).toBe(defaultProps.src)
    expect(imageElement?.getAttribute('alt')).toBe(defaultProps.alt)
    expect(imageElement?.className).contain('image-component')
    expect(container).toMatchSnapshot()
  })

  test('renders correctly with imgSet', () => {
    const { container } = render(<Image {...defaultProps} srcSet={imageSet} />)
    const imageElement = container.querySelector('img')
    expect(imageElement?.tagName).toBe('IMG')
    expect(imageElement?.getAttribute('src')).toBe(imageSet[0].src)
    expect(imageElement?.getAttribute('alt')).toBe(defaultProps.alt)
    expect(imageElement?.className).contain('image-component')
    expect(container).toMatchSnapshot()
  })

  test('renders correctly className', () => {
    const { container } = render(<Image {...defaultProps} />)
    const imageElement = container.querySelector('img')
    expect(imageElement?.className).contain(defaultProps.className)
    expect(container).toMatchSnapshot()
  })

  test('renders correctly disabled', () => {
    const { container } = render(<Image {...defaultProps} />)
    const imageElement = container.querySelector('img')
    expect(imageElement?.className).contain('disabled')
    expect(container).toMatchSnapshot()
  })
})
