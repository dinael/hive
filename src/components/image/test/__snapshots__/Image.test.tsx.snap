// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Image > renders correctly 1`] = `
<div>
  <img
    alt="Imagen description for accessibility"
    class="
          image-component
          disabled
          none
          cover
          className-test
        "
    height="300"
    loading="lazy"
    src="https://picsum.photos/1440/500?random=1"
    width="300"
  />
</div>
`;

exports[`Image > renders correctly className 1`] = `
<div>
  <img
    alt="Imagen description for accessibility"
    class="
          image-component
          disabled
          none
          cover
          className-test
        "
    height="300"
    loading="lazy"
    src="https://picsum.photos/1440/500?random=1"
    width="300"
  />
</div>
`;

exports[`Image > renders correctly disabled 1`] = `
<div>
  <img
    alt="Imagen description for accessibility"
    class="
          image-component
          disabled
          none
          cover
          className-test
        "
    height="300"
    loading="lazy"
    src="https://picsum.photos/1440/500?random=1"
    width="300"
  />
</div>
`;

exports[`Image > renders correctly with imgSet 1`] = `
<div>
  <img
    alt="Imagen description for accessibility"
    class="
        image-component
        disabled
        none
        cover
        className-test
      "
    height="300"
    loading="lazy"
    src="https://picsum.photos/1440/500?random=1"
    srcset="https://picsum.photos/1440/500?random=1 320w,https://picsum.photos/900/300?random=2 480w,https://picsum.photos/800/200?random=3 640w"
    width="300"
  />
</div>
`;

exports[`Image > renders snapshot 1`] = `
<div>
  <img
    alt="Imagen description for accessibility"
    class="
          image-component
          disabled
          none
          cover
          className-test
        "
    height="300"
    loading="lazy"
    src="https://picsum.photos/1440/500?random=1"
    width="300"
  />
</div>
`;
